File: library/src/main/java/com/github/easyguide/layer/GuideLayer.java
package com.github.easyguide.layer;

import android.app.Activity;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by shenxl on 2018/8/16.
 */

public class GuideLayer extends AbsGuideLayer
        implements GuideLayerView.DrawCallBack, GuideLayerView.LayerClickListener{
    private Activity mActivity;
    private GuideLayerView mViewContainer;
    private List<Rect> mTargetCache = new ArrayList<>();
    private List<View> mViewCache = new ArrayList<>();
    private onFullClickListener mFullClickListener;
    private onSingleClickListener mSingleClickListener;

    public GuideLayer(Activity activity) {
        mActivity = activity;
    }

    protected void onViewCreated(Context context){

    }

    protected Activity getActivity() {
        return mActivity;
    }

    public GuideLayerView getViewContainer() {
        return mViewContainer;
    }

    public final GuideLayer addTargetView(int id){
        return addTargetView(mActivity.findViewById(id));
    }

    public final GuideLayer addTargetView(final View view){
        view.post(new Runnable() {
            @Override
            public void run() {
                addTargetView(ViewLocUtils.getViewAbsRect(mActivity, view));
            }
        });
        return this;
    }

    public final GuideLayer addTargetView(Rect rect) {
        if (mViewContainer != null) {
            mViewContainer.addTargetsRect(rect);
            mViewContainer.requestLayout();
        } else {
            mTargetCache.add(rect);
        }
        return this;
    }

    public final GuideLayer addExtraView(View view, Location location, int tartgetIndex) {
        location.setIndex(tartgetIndex);
        view.setTag(location);
        if (mViewContainer != null) {
            mViewContainer.addView(view);
        } else {
            mViewCache.add(view);
        }
        return this;
    }

    public void setFullClickListener(onFullClickListener fullClickListener) {
        mFullClickListener = fullClickListener;
    }

    public void setSingleClickListener(onSingleClickListener singleClickListener) {
        mSingleClickListener = singleClickListener;
    }

    @Override
    public final View makeView(Context context) {
        mViewContainer = new GuideLayerView(context);

        mViewContainer.setDrawCallBack(this);
        mViewContainer.setLayerClickListener(this);
        for (int i = 0; i < mTargetCache.size(); i++) {
            addTargetView(mTargetCache.get(i));
        }
        for (int i = 0; i < mViewCache.size(); i++) {
            mViewContainer.addView(mViewCache.get(i));
        }
        onViewCreated(context);

        return mViewContainer;
    }

    @Override
    public void onDraw(int id, Rect rect, Canvas canvas, Paint paint){
        canvas.drawRect(rect, paint);
    }

    @Override
    public final void onFullClick() {
        if (mSingleClickListener == null) {
            if (mFullClickListener == null) {
                getCallback().dismissCurrent();
            } else {
                mFullClickListener.onClick(mViewContainer, getCallback());
            }
        }
    }

    @Override
    public final void onSingleClick(int id) {
        if (mSingleClickListener != null) {
            mSingleClickListener.onClick(id, mViewContainer, getCallback());
        }
    }

    public interface onFullClickListener{
        void onClick(GuideLayerView container, ILayerCallback callback);
    }

    public interface onSingleClickListener{
        void onClick(int id, GuideLayerView container, ILayerCallback callback);
    }
}

File: library/src/main/java/com/github/easyguide/layer/GuideLayer.kt
package com.github.easyguide.layer

import android.app.Activity
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.os.Build
import android.view.View
import com.github.easyguide.client.ILayerCallback

/**
 * Created by shenxl on 2018/8/16.
 */

open class GuideLayer(protected val activity: Activity) : AbsGuideLayer() {
    private lateinit var viewContainer: GuideLayerView
    private val targetCache = mutableListOf<Rect>()
    private val viewCache = mutableListOf<View>()
    var fullScreenClickListener: OnFullScreenClickListener? = null
    var singleTargetClickListener: OnSingleTargetClickListener? = null
    var defaultClickEnable = true

    protected open fun onViewCreated(context: Context) {
    }

    fun addTargetView(view: View): GuideLayer {
        view.post {
            addTargetView(getViewAbsRect(activity, view))
        }
        return this
    }

    fun addTargetView(rect: Rect): GuideLayer {
        if (this::viewContainer.isInitialized) {
            viewContainer.addTargetsRect(rect)
            viewContainer.requestLayout()
        } else {
            targetCache.add(rect)
        }
        return this
    }

    fun addExtraView(view: View, location: Location, tartgetIndex: Int): GuideLayer {
        location.setIndex(tartgetIndex)
        view.tag = location
        if (this::viewContainer.isInitialized) {
            viewContainer.addView(view)
        } else {
            viewCache.add(view)
        }
        return this
    }

    override fun makeView(context: Context): View {
        viewContainer = GuideLayerView(context).apply {
            drawCallBack = this@GuideLayer::onDraw
            totalClickListener = this@GuideLayer::onFullClick
            singleClickListener = this@GuideLayer::onSingleClick
            for (item in targetCache) {
                addTargetsRect(item)
            }
            for (item in viewCache) {
                addView(item)
            }
        }
        onViewCreated(context)
        return viewContainer
    }

    protected open fun onDraw(id: Int, rect: Rect, canvas: Canvas, paint: Paint) {
        canvas.drawRect(rect, paint)
    }

    private fun onFullClick() {
        if (singleTargetClickListener == null && defaultClickEnable) {
            fullScreenClickListener?.onClick(viewContainer, callback) ?: run {
                callback.dismissCurrent()
            }
        }
    }

    private fun onSingleClick(id: Int) {
        if (defaultClickEnable) {
            singleTargetClickListener?.onClick(id, viewContainer, callback)
        }
    }

    interface OnFullScreenClickListener {
        fun onClick(container: GuideLayerView, callback: ILayerCallback)
    }

    interface OnSingleTargetClickListener {
        fun onClick(id: Int, container: GuideLayerView, callback: ILayerCallback)
    }

    companion object {
        private fun getStatusBarHeight(context: Context): Int {
            val resourceId = context.resources.getIdentifier("status_bar_height", "dimen", "android")
            return context.resources.getDimensionPixelSize(resourceId)
        }

        fun getViewAbsRect(context: Context, view: View): Rect {
            val locView = IntArray(2)
            view.getLocationInWindow(locView)
            val rect = Rect()
            rect.set(locView[0], locView[1], locView[0] + view.measuredWidth, locView[1] + view.measuredHeight)
            rect.offset(0, if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) -getStatusBarHeight(context) else 0)
            return rect
        }
    }
}
